designs/2019-typescript-eslintrc-support/README.md

```diff
@@ -0,0 +1,351 @@
+
+By allowing first-class support for `.eslintrc.ts`, the amount of configuration
+problems & general confusion over such experienced by users can be reduced,
+since it would enable us to provide tooling to type configuration files and so
+allow TypeScript to catch ESLint configuration errors at compile time, rather
+than run time.
+
+While we can already hack ways to use `.eslintrc.ts`, none of them lend
+themselves to ease-of-use in the same manner that ESLint encourages, and so
+first-class support for `.eslintrc.ts` will go long way towards encouraging
+community driven typing of ESLint plugins.
+
+Finally, ESLint is being adopted as the linter for TypeScript, meaning
+first-class support for `.eslintrc.ts` will help enable a positive feedback loop
+to help improve the two tools.
+
+There are some potential UX benefits for users that use & understand TypeScript,
+in the form
+
+## Detailed Design
+
+- Adds support for `.eslintrc.ts`
+- Requires uses to install `ts-node` manually relative to the `node_modules` of
+  the configuration file, per standard Node
+  [module resolution logic](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders)
+  - Failure to do so with a `.eslintrc.ts` results in ESLint throwing a fatal
+    error
+  - The error message should include understandable plain text instructions on
+    how to fix the error.
+- ESLint does not validate the configuration object for now.
+- `.eslintrc.ts` is second highest priority, after `.eslintrc.js`
+  - This matches the priority of the other supported types, and generally where
+    TypeScript files it for other tools i.e webpack
+
+## Documentation
+
+I don't think a large amount of ceremony is required, but @typescript-eslint
+might like to do something.
+
+Overall, updating the list of accepted formats for `eslintrc`, and a small note
+with a link to @typescript-eslint I think would be all that's needed.
+
+@typescript-eslint probably would like to provide some documentation on how to
+nicely type & write definitions for ESLint plugins, but that's not strictly
+required by this feature.
+
+It should be documented that _only_ `.eslintrc.ts` is supported (i.e not
+`parsers`,`plugins`, etc), and that the shipping of `.ts` configurations is
+discouraged - instead they should be compiled into JavaScript, and shipped with
+a type definition file (`.eslintrc.d.ts`).
+
+## Drawbacks
+
+> Slight increase in maintenance burden for core support.
+
+People may open issues complaining about TypeScript-related errors, or errors
+relating to the registering of `ts-node`.
+
+The majority of these should be detectable from their stacktrace or use of
+keywords such as TypeScript, `ts`, `ts-node`, and `.eslintrc.ts`. As such, I'd
+expect it should be possible to refine `eslint-bot` to triage these issues and
+post a comment detailing common problems & their solutions to help reduce this
+cost.
+
+Personally I don't think the general amount of maintenance that would result
```

Comment:
Speaking from experience, I don't think it's wise to count on this kind of support. I think we can remove this paragraph from the RFC.

Reply:
Sure, but I do feel it's important to have this said, b/c there is a *whole core team* & community for eslinting TypeScript, which afaik has been doing so w/ no changes required to the core.

This is the first change we've hit a need for, and it's one that will enable us to continue to work independently to help improve the overall JS ecosystem.

I think it's fair to say that if the likes of maintainers like @bradzacher & JamesHenry disappear (and the rest of the amazing @typescript-eslint team), and the community drys up, the requirement to maintain types will also dwindle.

I completely agree not relying on it is wise, but I feel the above should be kept in mind b/c otherwise you could very easily decline this w/ the argument of "We'll have to do all the work always for any types relating to ESLint" 🙂